### PR TITLE
PI-3473: Netcdf saving of ancillary variables

### PR DESCRIPTION
**Pull request by pp-mo**
Saving ancillary variables, with some testcases based on CDL comparisons.

~There is some unfinished business here where the tests might be improved,~
~if and when they can use other aspects of ancillary support :~
**UPDATE:**
FIxed that.
Nearly good to go?

### CODE REVIEW

**Comment by pp-mo** [comment]
Hi @stephenworsley 
Now rebased, should be GTG ! :pray:

File: lib/iris/tests/test_netcdf.py [hunk 1]
```diff
@@ -964,6 +964,111 @@ def test_no_global_attributes(self):
...
+        testcube_1 = stock.realistic_3d()
+        ancil = iris.coords.AncillaryVariable(
+            np.zeros(testcube_1[0].shape), long_name="latlon_refs", units="s"
+        )
+        testcube_1.add_ancillary_variable(ancil, (1, 2))
+
+        testcube_2 = testcube_1.copy()
+        testcube_2.units = "m"
+        testcube_2.rename("alternate_data")
+        # TODO: this should not be necessary, when cube.copy() handles ancils.
```
Comment (stephenworsley) [inline lib/iris/tests/test_netcdf.py]: Now that #3546 is merged, does rebasing and modifying this code work as expected?

File: lib/iris/tests/test_netcdf.py [hunk 2]
```diff
@@ -964,6 +964,111 @@ def test_no_global_attributes(self):
...
+        testcube_2.rename("alternate_data")
+
+        ancil1 = iris.coords.AncillaryVariable(
+            np.zeros(testcube_1[0].shape), long_name="latlon_refs", units="s"
+        )
+        testcube_1.add_ancillary_variable(ancil1, (1, 2))
+
+        ancil2 = ancil1.copy()
+        ancil2.data[0, 0] += 1.0
+        # TODO: needs modifying, when cube.copy() handles ancils.
```
Comment (stephenworsley) [inline lib/iris/tests/test_netcdf.py]: As above, this should be possible to address now #3546 is merged.

File: lib/iris/fileformats/netcdf.py [hunk 3]
```diff
@@ -1093,14 +1093,18 @@ def write(
...
+        # Add the auxiliary coordinate variables and associate the data
         # variable to them
         self._add_aux_coords(cube, cf_var_cube, dimension_names)
 
-        # Add the cell_measures variable names and associate the data
+        # Add the cell_measures variables and associate the data
         # variable to them
         self._add_cell_measures(cube, cf_var_cube, dimension_names)
 
+        # Add the ancillary_variables variables and associate the data variable
```
[the remark this answers is not in the document]
Comment (pp-mo) [reply lib/iris/fileformats/netcdf.py]: "Variable names" is really _wrong_, I think.
What all these "_add" functions do is to is create variables in the datafile.

The line above just says "Add coordinate variables", which I think is the right way to describe it.

File: lib/iris/fileformats/netcdf.py [hunk 4]
```diff
@@ -1714,26 +1760,79 @@ def _create_cf_cell_measure_variable(
...
+        Returns:
+            The string name of the associated CF-netCDF variable saved.
+
+        """
+        # Note: currently shares variable creation code with cell-measures.
+        return self._inner_create_cf_cellmeasure_or_ancil_variable(
+            cube, dimension_names, ancillary_variable
+        )
+
+    # TODO: this was renamed from "_create_cf_variable" .
```
[the remark this answers is not in the document]
Comment (pp-mo) [reply lib/iris/fileformats/netcdf.py]: Not that I can think of.
Good spot! 😀

File: lib/iris/tests/test_netcdf.py (hunk 1 above)
Comment (pp-mo) [reply lib/iris/tests/test_netcdf.py]: Well, it probably *ought* to fail then, I think, complaining that the ancillary to be added is identical to one already there.
I will investigate  this ...

File: lib/iris/tests/test_netcdf.py (hunk 2 above)
Comment (pp-mo) [reply lib/iris/tests/test_netcdf.py]: :+1:

**Comment by pp-mo** [comment]
> it probably ought to fail then, I think, complaining that the ancillary to be added is identical to one already there.

Rebased, and it did fail as predicted :+1:
Now fixed.

File: lib/iris/tests/test_netcdf.py (hunk 2 above)
Comment (pp-mo) [reply lib/iris/tests/test_netcdf.py]: Note: this actually did *not* need code changing.
The cube is copied before an ancil is attached, so no collision between the two different versions occurs when adding to the second cube.
The two are stored separately (with disambiguated var_names), _because_ they contain different data.  
All good :+1: !

File: lib/iris/fileformats/netcdf.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (stephenworsley) [reply lib/iris/fileformats/netcdf.py]: Ah, okay I think I mixed up the red and green lines above here.

**Pull request merged**